src/ast_editor.ml

```diff
@@ -0,0 +1,532 @@
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  match fetch_pp_code ~document with
+  | exception Sys_error e -> Promise.return (Error e)
+  | pp_pp_str ->
+    let* doc =
+      Workspace.openTextDocument
+        (`Uri
+          (Uri.parse ("post-ppx: " ^ TextDocument.fileName document ^ "?") ()))
+    in
+    Ast_editor_state.set_changes_tracking ast_editor_state document doc;
+    replace_document_content ~content:pp_pp_str ~document:doc;
+    let+ (_ : TextEditor.t) =
+      Window.showTextDocument ~document:(`TextDocument doc)
+        ~column:ViewColumn.Beside ()
+    in
+    Ok 0
+
+let reload_pp_doc instance ~document =
+  let open Promise.Syntax in
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  let visibleTextEditors = Window.visibleTextEditors () in
+  let origin_uri =
+    match
+      Ast_editor_state.find_original_doc_by_pp_uri ast_editor_state
+        (TextDocument.uri document)
+    with
+    | Some x -> x
+    | None -> failwith "Failed finding the original document URI."
+  in
+  let* original_document =
+    Workspace.openTextDocument (`Uri (Uri.parse origin_uri ()))
+  in
+  match
+    List.find visibleTextEditors ~f:(fun editor ->
+        document_eq (TextEditor.document editor) document)
+  with
+  | None -> Promise.resolve 1
+  | Some _ ->
+    Ast_editor_state.set_origin_changed ast_editor_state
+      ~key:(TextDocument.uri document)
+      ~data:false;
+    replace_document_content
+      ~content:(fetch_pp_code ~document:original_document)
+      ~document;
+    Promise.resolve 0
+
+let rec manage_choice instance choice ~document : int Promise.t =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  match choice with
+  | Some `Update
+  | Some `Retry ->
+    (match
+       (Ast_editor_state.pp_status ast_editor_state) (TextDocument.uri document)
+     with
+    | `Original -> reload_pp_doc
+    | `Absent_or_pped -> open_preprocessed_doc_to_the_side)
+      instance ~document
+  | Some `Abandon -> Promise.resolve 1
+  | _ -> Promise.resolve (-1)
+
+and manage_open_failure err_msg instance ~document =
+  let open Promise.Syntax in
+  let* choice =
+    Window.showInformationMessage
+      ~message:(err_msg ^ " Please make sure your project is built and retry.")
+      ~choices:[ ("Retry", `Retry); ("Abandon", `Abandon) ]
+      ()
+  in
+  manage_choice instance choice ~document
+
+and open_preprocessed_doc_to_the_side instance ~document =
+  let open Promise.Syntax in
+  let* result = open_pp_doc instance ~document in
+  match result with
+  | Ok x -> Promise.return x
+  | Error e -> manage_open_failure e instance ~document
+
+let open_both_ppx_ast instance ~document =
+  let open Promise.Syntax in
+  let* pp_doc_open = open_preprocessed_doc_to_the_side instance ~document in
+  if pp_doc_open = 0 then
+    open_ast_explorer ~uri:(TextDocument.uri document)
+  else
+    Promise.make (fun ~resolve:_ ~reject:_ ->
```

Comment:
Why do we need this `Promise.make` wrapper? It should be removed here (and elsewhere).